### PR TITLE
Update Archipelago .net client to version 4.0

### PR DESCRIPTION
**Pull request by Dragonglove**
Primarily updated the mod to use the new Archipelago.MultiClient.Net version.  Much of the code remains in place, but a few minor changes have been made.

- CheckLocation now kicks off a separate task to send the packet to the Archipelago session.  This fixes a small hitch that would occur when collecting items.
- `ScoutCallback` logic moved around so it isn't a callback anymore.
- `SendPlacementHints` slightly modified.  May want to revisit the async/await logic here, but this seems to work for now.

Much of this is barely tested, and will need some full playthroughs be to confident for release.

### CODE REVIEW

File: Archipelago.HollowKnight/Archipelago.cs [hunk 1]
```diff
@@ -623,18 +623,21 @@ public void SendPlacementHints()
             LogDebug($"Hinting {hintedLocationIDs.Count()} locations.");
             try
             {
-                session.Socket.SendPacketAsync(new LocationScoutsPacket()
+                var locationScoutPacket = new LocationScoutsPacket()
```
Comment (dewiniaid) [inline Archipelago.HollowKnight/Archipelago.cs:626]: ~~We should update this whole chunk of code to use the createAsHint option available in `ScoutLocationsAsync`.  (That option didn't exist at the time I originally wrote this)~~

The remarks of ScoutLocationsAsync suggest this might actually be somewhat problematic if we were to send out a second scout before the first returns, so keep it as-is for now.

File: Archipelago.HollowKnight/Archipelago.cs [hunk 2]
```diff
@@ -597,12 +597,12 @@ public void SendPlacementHints()
             }
             HashSet<ArchipelagoItemTag> hintedTags = new();
             HashSet<long> hintedLocationIDs = new();
-            ArchipelagoItemTag tag;
```
Comment (dewiniaid) [inline Archipelago.HollowKnight/Archipelago.cs:600]: Why is this moved to be inside the loop?

**Review by dewiniaid — CHANGES_REQUESTED** [review]

File: Archipelago.HollowKnight/ArchipelagoRandomizer.cs [hunk 3]
```diff
@@ -140,23 +141,22 @@ public void Randomize()
             }
 
             // Scout all locations
-            void ScoutCallback(LocationInfoPacket packet)
+            var locations = new List<long>(session.Locations.AllLocations);
+            Task<LocationInfoPacket> packetTask = session.Locations.ScoutLocationsAsync(locations.ToArray());
+
+            LocationInfoPacket packet = packetTask.Result;
```
[the remark this answers is not in the document]
Comment (BadMagic100) [reply Archipelago.HollowKnight/ArchipelagoRandomizer.cs:147]: couple notes on this:

1. Nothing about this is actually waiting for the scout task to finish
2. Moving this into the BeginInvoke won't help, the point of that is that the stuff in that block *does run on the main thread* because unity doesn't like threading.

Unless I'm missing something it probably should be something like `ScoutLocationsAsync(...).ContinueWith((packetTask) => ThreadSupport.BeginInvoke(<do stuff with packetTask.result here>)`

**Comment by Ijwu** [comment]
I merged #130 which includes a custom merge job of me including the changes in this PR as well as #128. Thanks for kicking things off @Dragonglove even if it took a while to see action.
